daemon/algod/api/server/v2/handlers.go

```diff
@@ -182,6 +182,64 @@ func (v2 *Handlers) GetBlock(ctx echo.Context, round uint64, params generated.Ge
 	return ctx.Blob(http.StatusOK, contentType, data)
 }
 
+// GetProof generates a Merkle proof for a transaction in a block.
+// (GET /v2/blocks/{round}/transactions/{txid}/proof)
+func (v2 *Handlers) GetProof(ctx echo.Context, round uint64, txid string, params generated.GetProofParams) error {
+	var txID transactions.Txid
+	err := txID.UnmarshalText([]byte(txid))
+	if err != nil {
+		return badRequest(ctx, err, errNoTxnSpecified, v2.Log)
+	}
+
+	ledger := v2.Node.Ledger()
+	block, _, err := ledger.BlockCert(basics.Round(round))
+	if err != nil {
+		return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
+	}
+
+	proto := config.Consensus[block.CurrentProtocol]
+	if proto.PaysetCommit != config.PaysetCommitMerkle {
+		return notFound(ctx, err, "protocol does not support Merkle proofs", v2.Log)
+	}
+
+	txns, err := block.DecodePaysetFlat()
+	if err != nil {
+		return internalError(ctx, err, "decoding transactions", v2.Log)
+	}
+
+	for idx := range txns {
+		if txns[idx].Txn.ID() == txID {
+			tree, err := block.TxnMerkleTree()
+			if err != nil {
+				return internalError(ctx, err, "building Merkle tree", v2.Log)
+			}
+
+			proof, err := tree.Prove([]uint64{uint64(idx)})
+			if err != nil {
+				return internalError(ctx, err, "generating proof", v2.Log)
+			}
+
+			var proofconcat []byte
+			for _, proofelem := range proof {
+				proofconcat = append(proofconcat, proofelem[:]...)
+			}
```

Comment:
Did we benchmark these? Might be a fairly expensive group of operations for an API call.

Reply:
There are some benchmarks for various components of what's going on here (see various benchmarks in `crypto/merklearray`).  I think a rough upper bound would be `BenchmarkMerkleCommit/Item1000/Count10000` which takes 17 msec (although that's for a 10 MByte block containing 10K transactions of 1 KByte each).  Doesn't seem super expensive but certainly people operating publicly-accessible REST APIs (maybe Purestake?) should pay attention.

Reply:
Fair enough, 17ms is not a disaster for such a call.